libs/editor/egui_editor/src/apple/apple_ffi.rs

```diff
@@ -0,0 +1,123 @@
+use crate::apple::keyboard::NSKeys;
+use crate::{Editor, WgpuEditor};
+use egui::{Context, Event};
+use egui_wgpu_backend::{wgpu, ScreenDescriptor};
+use std::ffi::{c_char, c_void, CStr};
+
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn init_editor(
+    metal_layer: *mut c_void, //, width: u32, height: u32, scaling: u32,
+) -> *mut c_void {
+    let backend = wgpu::util::backend_bits_from_env().unwrap_or_else(wgpu::Backends::all);
+    let instance = wgpu::Instance::new(backend);
+    let surface = instance.create_surface_from_core_animation_layer(metal_layer);
+    let (adapter, device, queue) = pollster::block_on(request_device(&instance, backend, &surface));
+    let surface_format = surface.get_supported_formats(&adapter)[0];
+    let screen =
+        ScreenDescriptor { physical_width: 10000, physical_height: 10000, scale_factor: 1.0 };
+    let surface_config = wgpu::SurfaceConfiguration {
+        usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+        format: surface.get_supported_formats(&adapter)[0],
+        width: screen.physical_width, // TODO get from context or something
+        height: screen.physical_height,
+        present_mode: wgpu::PresentMode::Fifo,
+    };
+    surface.configure(&device, &surface_config);
+    let rpass = egui_wgpu_backend::RenderPass::new(&device, surface_format, 1);
+
+    let context = Context::default();
+    let editor = Editor::default();
+    editor.set_font(&context);
+
+    let mut obj = WgpuEditor {
+        device,
+        queue,
+        surface,
+        adapter,
+        rpass,
+        screen,
+        context,
+        raw_input: Default::default(),
+        editor,
+    };
+
+    obj.frame();
+
+    // TODO we need to free this memory
```

Comment:
Has this been taken care of yet?

Reply:
no, this todo will need to stay for a bit